### ISSUE TITLE
each token ends with a sace except the last one

### ISSUE THREAD

**Issue by RayRiding** [gurnec/btcrecover#109]
Firstly, this is an awesome tool, a fantastic piece of work.
For my electrum password, I decided to use part of a seed phrase. Therefore I know that each word (token) is followed by a space, except the final word.
The way I am doing this for now is to follow each token with %s wildcard to represent the space. Obviously, this won't be the case for the final word, so I have used --typos-delete (taking ages).
I noticed that I can use a typos map, but don't have a clue where to begin! 
Basically, what I'm hoping for is an instruction to only remove the space from any possible last words tried. Can this be done at all?

**Comment by RayRiding** [comment]
Thank you for the prompt and detailed response.
Unfortunately I cannot seem to get the first example to work. It just checks the number of passwords I have as required tokens or none!

The second example using the anchors I have had some success with, but the rate at which it checks has dropped from the usual 170 kP/s I get to just over 5kP/s. Would you have any idea as to why this is happening?

Also, I have found a superb tool for adding prefixes, suffixes, removing extra white space etc, which is perfect for these types of wordlists. It is called Text Mechanic [http://textmechanic.com/](url).
Well worth letting people know about it :)

**Comment by gurnec** [comment]
First off, you probably want to add `--max-tokens #` to set a max number of words/tokens to use. Keep in mind that the contracting wildcard counts as a token, so if you use `--max-tokens 5`, that will be 4 words plus the contracting wildcard.

The second example won't scale well if you're using more than around `--max-tokens 5`, I'd just ignore it and use the first example if you can.

> Unfortunately I cannot seem to get the first example to work. It just checks the number of passwords I have as required tokens or none!

I'm afraid I don't understand... could you post your tokens.txt file? (Obviously change the actual words to something different first!)

Also, excuse the dumb question, but if you have your seed can't you just restore from it instead?

**Comment by RayRiding** [comment]
I'm currently using --max-tokens 6, so that could explain that one.

Here is my tokens.txt file for the 1st example (I made a "test" wallet to make sure everything works)
+:cigar_
+:fashion_ 
fish_ 
critic_
wear_
plastic_ 
+:%-$

The _ is to show the invisible spaces
The password I am searching for is:
cigar fashion plastic

My command is as follows:
./btcrecover.py --wallet test_wallet --tokenlist tokens.txt --min-tokens 3 --max-tokens 3 --no-eta --no-dupcheck --delimiter :

As for the seed, this has been destroyed along with the password! That's another story!!!
Certainly appreciate the help - all I can do is promise a donation when I finally crack the password :)

**Comment by gurnec** [comment]
You have `--max-tokens 3`, and you have three required tokens (cigar, fashion, and the contracting wildcard), so the only passwords that it *can* generate that follow all the rules are "cigar fashion" and "fashion cigar" (both of which have 3 tokens, and it won't ever add more than the requested max). You'll want to up the `--max-tokens`.

If you only have 6 words, you may just want to get rid of `--max-tokens` entirely since there aren't that many combinations.

(Good idea testing this out on a test wallet by the way.)

**Comment by RayRiding** [comment]
Of course - forgot about the contracting wildcard being a token itself... All working now, fingers crossed again.
Thank you again for your expertise :)
I'm going to open another thread now lol

**Comment by nori19870522** [comment]
I am Japanese and do password analysis.
https://keychainx.io/ 

This is Robert's article.
https://news.bitstarz.com/bitstarz-news-interview-robert-rhodin-keychainx-ceo  

I may be able to find the password you lost. Do you need help with your login?

You can contact us on https://keychainx.io or send mail to help@keychainx.io
